### PR TITLE
[InstCombine] simplify `(X * C0) / (X * C1)` into `C0 / C1`.

### LINKED ISSUE #72114: [InstCombine] (m+m)/(m) is not optimized when m=a+a+a, but it is for m=a+a
https://godbolt.org/z/Y84oh497E

```cpp
int m, n;
void func(int a){

    m=a+a+a;
    n=(m+m)/(m);
}
```
We note that the expression `(m+m)/m` is optimized to `2` if the number of a in the expression assigned to m is 2ⁿ, otherwise it is not optimized.

proof:
https://alive2.llvm.org/ce/z/y9amEc

### PR DESCRIPTION
**Pull request by Z572**
fix #72114
proof: https://alive2.llvm.org/ce/z/xqprFm

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1214]: We should also check the `nsw` flag for `Op1`.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1215]: `udiv (X *nuw Y), (X *nuw Z) -> udiv Y, Z` has been handled.
https://github.com/llvm/llvm-project/blob/dc683d2e66de79bbea786f51788961eec5d0b793/llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp#L1380-L1393

Could you please remove this and implement the fold for both `sdiv` and `udiv` in `commonIDivTransforms`?

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1219]: Please use `BinaryOperator::CreateUDiv` and `BinaryOperator::CreateSDiv`.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1212]: I think we can generalize this pattern to handle `sdiv (X *nsw Y), (X *nsw Z) -> sdiv Y, Z` iff `sdiv Y, Z` doesn't overflow.
Alive2: https://alive2.llvm.org/ce/z/5TdmfG

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (Z572) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp]: What do I need to do

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp]: > What do I need to do

You should check `c1 != -1 || c1 != sign mask`.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1234]: I think it may cause miscompilation. I believe it is safe iff `A / B` is exact.
Please provide alive2 proof for it.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1241]:
```suggestion
      if (auto *Val = CreateDivOrNull(Y, Z))
        return Val;
```

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1246]:
```suggestion
      if (auto *Val = CreateDivOrNull(X, Z))
        return Val;
```

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1222]: Sorry for my mistake here.
You need to check `A != sign mask || B != -1`.
Since `mul nsw X, -1` will be canonicalized into `sub nsw 0, X`, I think the defense checking `B != -1` is enough as @goldsteinn suggested.

Please also add the following test:
```
define i4 @sdiv_mul_nsw_mul_nsw_allones(i4 %a, i4 %c2) {
  %add4 = mul nsw i4 %a, -1
  %add5 = mul nsw i4 %a, -8
  %div = sdiv i4 %add5, %add4
  ret i4 %div
}
```

File: llvm/test/Transforms/InstCombine/div.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/div.ll:1647]:
```suggestion
  %add4 = mul nsw i4 %a, -8
```

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1234]: Alive2: https://alive2.llvm.org/ce/z/2gkqu7

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1223]: Looks like the new version of your patch cannot handle the following pattern:
```
define i4 @src(i4 %a) {
  %add4 = mul i4 %a, 3
  %add5 = mul nuw i4 %a, 6
  %div = udiv i4 %add5, %add4
  ret i4 %div
}

define i4 @tgt(i4 %a) {
  ret i4 2
}
```
Alive2: https://alive2.llvm.org/ce/z/27Ei4q

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1226]: Possible implementation:
```suggestion
      if (!IsSigned && OB0HasNUW) {
         if (auto *CV = dyn_cast<Constant>(simplifyUDivInst(A, B, /*IsExact*/ true, SQ.getWithInstruction(&I)));
         CV && !(isa<PoisonValue>(CV) || CV->containsPoisonElement()))
             return CV;
      }
      return nullptr;
```

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1220]: It also holds without the `nsw` flag on OB1 if `A` is a multiple of `B`.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (Z572) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp]: Maybe this is better? https://alive2.llvm.org/ce/z/HL3K6G

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1221]: `*C != -1` -> `!C->isAllOnes()`?

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1229]:
```suggestion
        if (match(A, m_APInt(C1)) && match(B, m_APInt(C2)) && !C2->isSignMask()) {
          APInt Q, R;
          APInt::sdivrem(*C1, *C2, Q, R);
          if (R->isZero())
            return replaceInstUsesWith(I, ConstantInt::get(A->getType(), Q));
        }
```
I think this implementation is simpler.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1238]:
```suggestion
        if (match(A, m_APInt(C1)) && match(B, m_APInt(C2))) {
          APInt Q, R;
          APInt::udivrem(*C1, *C2, Q, R);
          if (R->isZero())
            return replaceInstUsesWith(I, ConstantInt::get(A->getType(), Q));
        }
```

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1246]:
```suggestion
    }
```

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1221]:
```suggestion
        if (OB1HasNSW && match(B, m_APInt(C)) && !C->isAllOnes())
```

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1210]:
```suggestion
  // (X * Y) / (X * Z) --> Y / Z (and commuted variants)
```

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1237]: Wouldn't it be sufficient to check that `C2 <= C1` here? https://alive2.llvm.org/ce/z/B8cFaS

(Sorry if this was already discussed, I didn't follow the review.)

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1224]: This `!C2->isSignMask()` check doesn't seem to match what the alive2 proof does, which checks whether the multiplication result is the sign mask, not one of the multiplication operands.

It does not verify if I check `%c1` instead: https://alive2.llvm.org/ce/z/5uSr3w (It's pretty confusing that the meaning of c1/c2 in the proofs and C1/C2 in the code is inverted.)

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1237]: You are right.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1224]: Sorry for my mistake :(
I have swapped c1 and c2 in the proof to match the code.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1231]: I don't think it is worth handling this pattern :(
```suggestion
```

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1236]: Could you please add a negative test?
```
define i4 @udiv_mul_nuw_mul_negative(i4 %a) {
  %add4 = mul i4 %a, 13
  %add5 = mul nuw i4 %a, 4
  %div = udiv i4 %add5, %add4
  ret i4 %div
}

```
Alive2: https://alive2.llvm.org/ce/z/K9Fann

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for additional approval from other reviewers.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**
